File: 2022/ShortDeal/app/src/main/java/com/creator/shortdeal/ImageSelectActivity.java
package com.creator.shortdeal;

import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.database.ContentObserver;
import android.database.Cursor;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.provider.MediaStore;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.ActionMode;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.GridView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.creator.shortdeal.adapters.CustomImageSelectAdapter;
import com.creator.shortdeal.common.HNApplication;
import com.creator.shortdeal.helpers.Constants;
import com.creator.shortdeal.models.Image;

import java.io.File;
import java.util.ArrayList;
import java.util.HashSet;

/**
 * Created by Darshan on 4/18/2015.
 */
public class ImageSelectActivity extends HelperActivity {
    private ArrayList<Image> images;
    private String album;

    private TextView errorDisplay;
    private TextView imageCount;

    private ProgressBar progressBar;
    private GridView gridView;
    private CustomImageSelectAdapter adapter;

    private ActionBar actionBar;

    private ActionMode actionMode;
    private int countSelected;

    private ContentObserver observer;
    private Handler handler;
    private Thread thread;

    private final String[] projection = new String[]{ MediaStore.Images.Media._ID, MediaStore.Images.Media.DISPLAY_NAME, MediaStore.Images.Media.DATA };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_image_select);
        setView(findViewById(R.id.layout_image_select));

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        Intent intent = getIntent();
        if (intent == null) {
            finish();
        }
        album = intent.getStringExtra(Constants.INTENT_EXTRA_ALBUM);

        actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowHomeEnabled(true);

            actionBar.setDisplayShowTitleEnabled(true);
            actionBar.setTitle(album);

            TextView textview_header_right = (TextView)findViewById(R.id.textview_header_right);
            textview_header_right.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    sendIntent();
                }
            });
        }

        imageCount = (TextView)findViewById(R.id.textview_header_image_count);

        errorDisplay = (TextView) findViewById(R.id.text_view_error);
        errorDisplay.setVisibility(View.INVISIBLE);

        progressBar = (ProgressBar) findViewById(R.id.progress_bar_image_select);
        gridView = (GridView) findViewById(R.id.grid_view_image_select);
        gridView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
//                if (actionMode == null) {
//                    actionMode = ImageSelectActivity.this.startActionMode(callback);
//                }
                toggleSelection(position);
//                if (countSelected == 0) {
//                    actionMode.finish();
//                }
                imageCount.setText(countSelected + "/" + HNApplication.LIMIT_IMAGE_COUNT);
            }
        });
    }

    @Override
    protected void onStart() {
        super.onStart();

        handler = new Handler() {
            @Override
            public void handleMessage(Message msg) {
                switch (msg.what) {
                    case Constants.PERMISSION_GRANTED: {
                        loadImages();
                        break;
                    }

                    case Constants.FETCH_STARTED: {
                        progressBar.setVisibility(View.VISIBLE);
                        gridView.setVisibility(View.INVISIBLE);
                        break;
                    }

                    case Constants.FETCH_COMPLETED: {
                        /*
                        If adapter is null, this implies that the loaded images will be shown
                        for the first time, hence send FETCH_COMPLETED message.
                        However, if adapter has been initialised, this thread was run either
                        due to the activity being restarted or content being changed.
                         */
                        if (adapter == null) {
                            adapter = new CustomImageSelectAdapter(getApplicationContext(), images);
                            gridView.setAdapter(adapter);

                            progressBar.setVisibility(View.INVISIBLE);
                            gridView.setVisibility(View.VISIBLE);
                            orientationBasedUI(getResources().getConfiguration().orientation);

                        } else {
                            adapter.notifyDataSetChanged();
                            /*
                            Some selected images may have been deleted
                            hence update action mode title
                             */
                            if (actionMode != null) {
                                countSelected = msg.arg1;
                                actionMode.setTitle(countSelected + " " + getString(R.string.selected));
                            }
                        }
                        break;
                    }

                    case Constants.ERROR: {
                        progressBar.setVisibility(View.INVISIBLE);
                        errorDisplay.setVisibility(View.VISIBLE);
                        break;
                    }

                    default: {
                        super.handleMessage(msg);
                    }
                }
            }
        };
        observer = new ContentObserver(handler) {
            @Override
            public void onChange(boolean selfChange) {
                loadImages();
            }
        };
        getContentResolver().registerContentObserver(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, false, observer);

        checkPermission();
    }

    @Override
    protected void onStop() {
        super.onStop();

        stopThread();

        getContentResolver().unregisterContentObserver(observer);
        observer = null;

        if (handler != null) {
            handler.removeCallbacksAndMessages(null);
            handler = null;
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (actionBar != null) {
            actionBar.setHomeAsUpIndicator(null);
        }
        images = null;
        if (adapter != null) {
            adapter.releaseResources();
        }
        gridView.setOnItemClickListener(null);
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        orientationBasedUI(newConfig.orientation);
    }

    private void orientationBasedUI(int orientation) {
        final WindowManager windowManager = (WindowManager) getApplicationContext().getSystemService(Context.WINDOW_SERVICE);
        final DisplayMetrics metrics = new DisplayMetrics();
        windowManager.getDefaultDisplay().getMetrics(metrics);

        if (adapter != null) {
            int size = orientation == Configuration.ORIENTATION_PORTRAIT ? metrics.widthPixels / 3 : metrics.widthPixels / 5;
            adapter.setLayoutParams(size);
        }
        gridView.setNumColumns(orientation == Configuration.ORIENTATION_PORTRAIT ? 3 : 5);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home: {
                onBackPressed();
                return true;
            }

            default: {
                return false;
            }
        }
    }

    private ActionMode.Callback callback = new ActionMode.Callback() {
        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            MenuInflater menuInflater = mode.getMenuInflater();
            menuInflater.inflate(R.menu.menu_contextual_action_bar, menu);

            actionMode = mode;
            countSelected = 0;

            return true;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return false;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {int i = item.getItemId();
            if (i == R.id.menu_item_add_image) {
                sendIntent();
                return true;
            }
            return false;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            if (countSelected > 0) {
                deselectAll();
            }
            actionMode = null;
        }
    };

    private void toggleSelection(int position) {
        if (!images.get(position).isSelected && countSelected >= Constants.limit) {
//            Toast.makeText(
//                    getApplicationContext(),
//                    String.format(getString(R.string.limit_exceeded), Constants.limit),
//                    Toast.LENGTH_SHORT)
//                    .show();
            return;
        }

        images.get(position).isSelected = !images.get(position).isSelected;
        if (images.get(position).isSelected) {
            countSelected++;

            images.get(position).sequence = countSelected;
//            Log.d("===============", images.get(position).sequence + "//POSITION");
        } else {
            countSelected--;

            // 재정렬
            int tmpSequence = images.get(position).sequence;
            images.get(position).sequence = -1;
//            Log.d("delete===============", "toggleSelection = " + position + "// sequence = " + images.get(position).sequence);
            for (int idx = 0; idx < images.size(); idx++) {
                if (images.get(idx).isSelected) {
                    if(images.get(idx).sequence > tmpSequence) {
                        images.get(idx).sequence = images.get(idx).sequence - 1;
                    }
                }
//                Log.d("update===============", "idx = " + idx + " // sequence = " + images.get(idx).sequence);
            }
        }

        adapter.notifyDataSetChanged();
    }

    private void deselectAll() {
        for (int i = 0, l = images.size(); i < l; i++) {
            images.get(i).isSelected = false;
        }
        countSelected = 0;
        adapter.notifyDataSetChanged();
    }

    private ArrayList<Image> getSelected() {
        String path = "";

        ArrayList<Image> selectedImages = new ArrayList<>(countSelected);
        selectedImages.clear();
        for (int i = 0; i < countSelected; i++) {
            selectedImages.add(i, images.get(0));
        }
        Log.d("SeongKwon", selectedImages.size() + "//" + countSelected);

        for (int i = 0; i < images.size(); i++) {
            if (images.get(i).isSelected) {
                path = path + ((images.get(i).sequence - 1)) + "//" + images.get(i).path + "\n\n";
                Log.d("SeongKwon", "selected path = " + images.get(i).path);
                Log.d("SeongKwon", (images.get(i).sequence - 1) + "//" + selectedImages.size() + "//" + countSelected);
                selectedImages.set(images.get(i).sequence - 1, images.get(i));
            }
        }

        Log.d("SeongKwon", "countSelected = " + countSelected);
        Log.d("SeongKwon", "images.size() = " + images.size());
        Log.d("SeongKwon", "selectedImages.size() = " + selectedImages.size());

        return selectedImages;
    }

    private void sendIntent() {
        Intent intent = new Intent();
        intent.putParcelableArrayListExtra(Constants.INTENT_EXTRA_IMAGES, getSelected());
        setResult(RESULT_OK, intent);
        ImageSelectActivity.this.finish();
    }

    private void loadImages() {
        startThread(new ImageLoaderRunnable());
    }

    private class ImageLoaderRunnable implements Runnable {
        @Override
        public void run() {
//            android.os.Process.setThreadPriority(java.lang.Process.THREAD_PRIORITY_BACKGROUND);
            /*
            If the adapter is null, this is first time this activity's view is
            being shown, hence send FETCH_STARTED message to show progress bar
            while images are loaded from phone
             */
            if (adapter == null) {
                sendMessage(Constants.FETCH_STARTED);
            }

            File file;
            HashSet<Long> selectedImages = new HashSet<>();
            if (images != null) {
                Image image;
                for (int i = 0, l = images.size(); i < l; i++) {
                    image = images.get(i);
                    file = new File(image.path);
                    if (file.exists() && image.isSelected) {
                        selectedImages.add(image.id);
                    }
                }
            }

            Cursor cursor = null;
            if(album.equals("전체 보기")) {
                cursor = getContentResolver().query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, projection, null, null, MediaStore.Images.Media.DATE_ADDED);
            } else {
                cursor = getContentResolver().query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, projection,
                        MediaStore.Images.Media.BUCKET_DISPLAY_NAME + " =?", new String[]{ album }, MediaStore.Images.Media.DATE_ADDED);
            }

            if (cursor == null) {
                sendMessage(Constants.ERROR);
                return;
            }

            /*
            In case this runnable is executed to onChange calling loadImages,
            using countSelected variable can result in a race condition. To avoid that,
            tempCountSelected keeps track of number of selected images. On handling
            FETCH_COMPLETED message, countSelected is assigned value of tempCountSelected.
             */
            int tempCountSelected = 0;
            ArrayList<Image> temp = new ArrayList<>(cursor.getCount());
            if (cursor.moveToLast()) {
                do {
                    if (Thread.interrupted()) {
                        return;
                    }

                    long id = cursor.getLong(cursor.getColumnIndex(projection[0]));
                    String name = cursor.getString(cursor.getColumnIndex(projection[1]));
                    String path = cursor.getString(cursor.getColumnIndex(projection[2]));
                    boolean isSelected = selectedImages.contains(id);
                    if (isSelected) {
                        tempCountSelected++;
                    }

                    file = new File(path);
                    if (file.exists()) {
//                        temp.add(new Image(id, name, path, isSelected));
                        temp.add(new Image(id, name, path, isSelected, -1));
                    }

                } while (cursor.moveToPrevious());
            }
            cursor.close();

            if (images == null) {
                images = new ArrayList<>();
            }
            images.clear();
            images.addAll(temp);

            sendMessage(Constants.FETCH_COMPLETED, tempCountSelected);
        }
    }

    private void startThread(Runnable runnable) {
        stopThread();
        thread = new Thread(runnable);
        thread.start();
    }

    private void stopThread() {
        if (thread == null || !thread.isAlive()) {
            return;
        }

        thread.interrupt();
        try {
            thread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private void sendMessage(int what) {
        sendMessage(what, 0);
    }

    private void sendMessage(int what, int arg1) {
        if (handler == null) {
            return;
        }

        Message message = handler.obtainMessage();
        message.what = what;
        message.arg1 = arg1;
        message.sendToTarget();
    }

    @Override
    protected void permissionGranted() {
        sendMessage(Constants.PERMISSION_GRANTED);
    }

    @Override
    protected void hideViews() {
        progressBar.setVisibility(View.INVISIBLE);
        gridView.setVisibility(View.INVISIBLE);
    }
}

File: 2024/ConceptK/app/src/main/java/com/creator/conceptk/ImageSelectActivity.kt
package com.creator.conceptk

import android.content.Intent
import android.content.res.Configuration
import android.database.ContentObserver
import android.database.Cursor
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.provider.MediaStore
import android.util.DisplayMetrics
import android.util.Log
import android.view.*
import android.widget.AdapterView.OnItemClickListener
import android.widget.GridView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.ActionBar
import androidx.appcompat.widget.Toolbar
import com.creator.conceptk.adapters.CustomImageSelectAdapter
import com.creator.conceptk.common.HNApplication
import com.creator.conceptk.helpers.Constants
import com.creator.conceptk.models.Image
import java.io.File

/**
 * Created by Darshan on 4/18/2015.
 */
class ImageSelectActivity : HelperActivity() {
    private var images: ArrayList<Image>? = null
    private var album: String? = null
    private var errorDisplay: TextView? = null
    private var imageCount: TextView? = null
    private var progressBar: ProgressBar? = null
    private var gridView: GridView? = null
    private var adapter: CustomImageSelectAdapter? = null
    private var actionBar: ActionBar? = null
    private var actionMode: ActionMode? = null
    private var countSelected = 0
    private var observer: ContentObserver? = null
    private var handler: Handler? = null
    private var thread: Thread? = null
    private val projection = arrayOf(
        MediaStore.Images.Media._ID,
        MediaStore.Images.Media.DISPLAY_NAME,
        MediaStore.Images.Media.DATA
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_image_select)
        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        val intent = intent
        if (intent == null) {
            finish()
        }
        album = intent!!.getStringExtra(Constants.INTENT_EXTRA_ALBUM)
        actionBar = supportActionBar
        if (actionBar != null) {
            actionBar!!.setDisplayHomeAsUpEnabled(true)
            actionBar!!.setDisplayShowHomeEnabled(true)
            actionBar!!.setDisplayShowTitleEnabled(true)
            actionBar!!.setTitle(album)
            val textview_header_right = findViewById<View>(R.id.textview_header_right) as TextView
            textview_header_right.setOnClickListener { sendIntent() }
        }
        imageCount = findViewById<View>(R.id.textview_header_image_count) as TextView
        errorDisplay = findViewById<View>(R.id.text_view_error) as TextView
        errorDisplay!!.visibility = View.INVISIBLE
        progressBar = findViewById<View>(R.id.progress_bar_image_select) as ProgressBar
        gridView = findViewById<View>(R.id.grid_view_image_select) as GridView
        gridView!!.onItemClickListener =
            OnItemClickListener { parent, view, position, id -> //                if (actionMode == null) {
//                    actionMode = ImageSelectActivity.this.startActionMode(callback);
//                }
                toggleSelection(position)
                //                if (countSelected == 0) {
//                    actionMode.finish();
//                }
                imageCount!!.text =
                    countSelected.toString() + "/" + HNApplication.Companion.LIMIT_IMAGE_COUNT
            }
    }

    override fun onStart() {
        super.onStart()
        handler = object : Handler() {
            override fun handleMessage(msg: Message) {
                when (msg.what) {
                    Constants.PERMISSION_GRANTED -> {
                        loadImages()
                    }
                    Constants.FETCH_STARTED -> {
                        progressBar!!.visibility = View.VISIBLE
                        gridView!!.visibility = View.INVISIBLE
                    }
                    Constants.FETCH_COMPLETED -> {

                        /*
                        If adapter is null, this implies that the loaded images will be shown
                        for the first time, hence send FETCH_COMPLETED message.
                        However, if adapter has been initialised, this thread was run either
                        due to the activity being restarted or content being changed.
                         */if (adapter == null) {
                            adapter = CustomImageSelectAdapter(applicationContext, images)
                            gridView!!.adapter = adapter
                            progressBar!!.visibility = View.INVISIBLE
                            gridView!!.visibility = View.VISIBLE
                            orientationBasedUI(resources.configuration.orientation)
                        } else {
                            adapter!!.notifyDataSetChanged()
                            /*
                            Some selected images may have been deleted
                            hence update action mode title
                             */if (actionMode != null) {
                                countSelected = msg.arg1
                                actionMode!!.title =
                                    countSelected.toString() + " " + getString(R.string.selected)
                            }
                        }
                    }
                    Constants.ERROR -> {
                        progressBar!!.visibility = View.INVISIBLE
                        errorDisplay!!.visibility = View.VISIBLE
                    }
                    else -> {
                        super.handleMessage(msg)
                    }
                }
            }
        }
        observer = object : ContentObserver(handler) {
            override fun onChange(selfChange: Boolean) {
                loadImages()
            }
        }
        contentResolver.registerContentObserver(
            MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
            false,
            observer!!
        )
        checkPermission()
    }

    override fun onStop() {
        super.onStop()
        stopThread()
        contentResolver.unregisterContentObserver(observer!!)
        observer = null
        if (handler != null) {
            handler!!.removeCallbacksAndMessages(null)
            handler = null
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (actionBar != null) {
            actionBar!!.setHomeAsUpIndicator(null)
        }
        images = null
        if (adapter != null) {
            adapter!!.releaseResources()
        }
        gridView!!.onItemClickListener = null
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        orientationBasedUI(newConfig.orientation)
    }

    private fun orientationBasedUI(orientation: Int) {
        val windowManager = applicationContext.getSystemService(WINDOW_SERVICE) as WindowManager
        val metrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(metrics)
        if (adapter != null) {
            val size =
                if (orientation == Configuration.ORIENTATION_PORTRAIT) metrics.widthPixels / 3 else metrics.widthPixels / 5
            adapter!!.setLayoutParams(size)
        }
        gridView!!.numColumns = if (orientation == Configuration.ORIENTATION_PORTRAIT) 3 else 5
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                true
            }
            else -> {
                false
            }
        }
    }

    private val callback: ActionMode.Callback = object : ActionMode.Callback {
        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            val menuInflater = mode.menuInflater
            menuInflater.inflate(R.menu.menu_contextual_action_bar, menu)
            actionMode = mode
            countSelected = 0
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
            return false
        }

        override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
            val i = item.itemId
            if (i == R.id.menu_item_add_image) {
                sendIntent()
                return true
            }
            return false
        }

        override fun onDestroyActionMode(mode: ActionMode) {
            if (countSelected > 0) {
                deselectAll()
            }
            actionMode = null
        }
    }

    private fun toggleSelection(position: Int) {
        if (!images!![position].isSelected && countSelected >= Constants.limit) {
//            Toast.makeText(
//                    getApplicationContext(),
//                    String.format(getString(R.string.limit_exceeded), Constants.limit),
//                    Toast.LENGTH_SHORT)
//                    .show();
            return
        }
        images!![position].isSelected = !images!![position].isSelected
        if (images!![position].isSelected) {
            countSelected++
            images!![position].sequence = countSelected
            //            Log.d("===============", images.get(position).sequence + "//POSITION");
        } else {
            countSelected--

            // 재정렬
            val tmpSequence = images!![position].sequence
            images!![position].sequence = -1
            //            Log.d("delete===============", "toggleSelection = " + position + "// sequence = " + images.get(position).sequence);
            for (idx in images!!.indices) {
                if (images!![idx].isSelected) {
                    if (images!![idx].sequence > tmpSequence) {
                        images!![idx].sequence = images!![idx].sequence - 1
                    }
                }
                //                Log.d("update===============", "idx = " + idx + " // sequence = " + images.get(idx).sequence);
            }
        }
        adapter!!.notifyDataSetChanged()
    }

    private fun deselectAll() {
        var i = 0
        val l = images!!.size
        while (i < l) {
            images!![i].isSelected = false
            i++
        }
        countSelected = 0
        adapter!!.notifyDataSetChanged()
    }

    private val selected: ArrayList<Image>
        private get() {
            var path = ""
            val selectedImages = ArrayList<Image>(countSelected)
            selectedImages.clear()
            for (i in 0 until countSelected) {
                selectedImages.add(i, images!![0])
            }
            Log.d("SeongKwon", selectedImages.size.toString() + "//" + countSelected)
            for (i in images!!.indices) {
                if (images!![i].isSelected) {
                    path = """
                    $path${images!![i].sequence - 1}//${images!![i].path}
                    
                    
                    """.trimIndent()
                    Log.d("SeongKwon", "selected path = " + images!![i].path)
                    Log.d(
                        "SeongKwon",
                        (images!![i].sequence - 1).toString() + "//" + selectedImages.size + "//" + countSelected
                    )
                    selectedImages[images!![i].sequence - 1] = images!![i]
                }
            }
            Log.d("SeongKwon", "countSelected = $countSelected")
            Log.d("SeongKwon", "images.size() = " + images!!.size)
            Log.d("SeongKwon", "selectedImages.size() = " + selectedImages.size)
            return selectedImages
        }

    private fun sendIntent() {
        val intent = Intent()
        intent.putParcelableArrayListExtra(Constants.INTENT_EXTRA_IMAGES, selected)
        setResult(RESULT_OK, intent)
        finish()
    }

    private fun loadImages() {
        startThread(ImageLoaderRunnable())
    }

    private inner class ImageLoaderRunnable : Runnable {
        override fun run() {
//            android.os.Process.setThreadPriority(java.lang.Process.THREAD_PRIORITY_BACKGROUND);
            /*
            If the adapter is null, this is first time this activity's view is
            being shown, hence send FETCH_STARTED message to show progress bar
            while images are loaded from phone
             */
            if (adapter == null) {
                sendMessage(Constants.FETCH_STARTED)
            }
            var file: File
            val selectedImages = HashSet<Long>()
            if (images != null) {
                var image: Image
                var i = 0
                val l = images!!.size
                while (i < l) {
                    image = images!![i]
                    file = File(image.path)
                    if (file.exists() && image.isSelected) {
                        selectedImages.add(image.id)
                    }
                    i++
                }
            }
            var cursor: Cursor? = null
            cursor = if (album == "전체 보기") {
                contentResolver.query(
                    MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                    projection,
                    null,
                    null,
                    MediaStore.Images.Media.DATE_ADDED
                )
            } else {
                contentResolver.query(
                    MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                    projection,
                    MediaStore.Images.Media.BUCKET_DISPLAY_NAME + " =?",
                    arrayOf(album),
                    MediaStore.Images.Media.DATE_ADDED
                )
            }
            if (cursor == null) {
                sendMessage(Constants.ERROR)
                return
            }

            /*
            In case this runnable is executed to onChange calling loadImages,
            using countSelected variable can result in a race condition. To avoid that,
            tempCountSelected keeps track of number of selected images. On handling
            FETCH_COMPLETED message, countSelected is assigned value of tempCountSelected.
             */
            var tempCountSelected = 0
            val temp = ArrayList<Image>(cursor.count)
            if (cursor.moveToLast()) {
                do {
                    if (Thread.interrupted()) {
                        return
                    }
                    val id = cursor.getLong(cursor.getColumnIndex(projection[0]))
                    val name = cursor.getString(cursor.getColumnIndex(projection[1]))
                    val path = cursor.getString(cursor.getColumnIndex(projection[2]))
                    val isSelected = selectedImages.contains(id)
                    if (isSelected) {
                        tempCountSelected++
                    }
                    file = File(path)
                    if (file.exists()) {
//                        temp.add(new Image(id, name, path, isSelected));
                        temp.add(Image(id, name, path, isSelected, -1))
                    }
                } while (cursor.moveToPrevious())
            }
            cursor.close()
            if (images == null) {
                images = ArrayList()
            }
            images!!.clear()
            images!!.addAll(temp)
            sendMessage(Constants.FETCH_COMPLETED, tempCountSelected)
        }
    }

    private fun startThread(runnable: Runnable) {
        stopThread()
        thread = Thread(runnable)
        thread!!.start()
    }

    private fun stopThread() {
        if (thread == null || !thread!!.isAlive) {
            return
        }
        thread!!.interrupt()
        try {
            thread!!.join()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    private fun sendMessage(what: Int, arg1: Int = 0) {
        if (handler == null) {
            return
        }
        val message = handler!!.obtainMessage()
        message.what = what
        message.arg1 = arg1
        message.sendToTarget()
    }

    override fun permissionGranted() {
        sendMessage(Constants.PERMISSION_GRANTED)
    }

    override fun hideViews() {
        progressBar!!.visibility = View.INVISIBLE
        gridView!!.visibility = View.INVISIBLE
    }
}